chainimport/headers_import_test.go

```diff
@@ -0,0 +1,8333 @@
+
+	// Create a temporary file.
+	fileName := fmt.Sprintf("test-%s-*", hT)
+	tempFile, err := os.CreateTemp("", fileName)
+	cleanup := func() {
+		tempFile.Close()
+		os.Remove(tempFile.Name())
+	}
+	if err != nil {
+		return nil, cleanup, err
+	}
+
+	var hdrsData []string
+
+	switch hT {
+	case headerfs.Block:
+		hdrsData = blockHdrs
+	case headerfs.RegularFilter:
+		hdrsData = filterHdrs
+	default:
+		return nil, cleanup, fmt.Errorf("%s", hT)
+	}
+
+	if include_metadata {
+		// Utilize AddHeadersImportMetadata func part of the chainimport
+		// package.
+		err = AddHeadersImportMetadata(
+			tempFile.Name(), wire.SimNet, hT, 0,
+		)
+		if err != nil {
+			return nil, cleanup, err
+		}
+
+		// We need to reopen it again to update the file descriptor
+		// since AddHeadersImportMetadata closes the file atomically.
+		tempFile, err = os.OpenFile(
+			tempFile.Name(), os.O_RDWR|os.O_APPEND, 0644,
+		)
+		if err != nil {
+			return nil, cleanup, fmt.Errorf("failed to open file "+
+				"for writing headers: %w", err)
+		}
+	}
+
+	// Now append header data to the file. Write each header as raw bytes.
+	for _, hdrHex := range hdrsData {
+		hdrBytes, err := hex.DecodeString(hdrHex)
+		if err != nil {
+			return nil, cleanup, fmt.Errorf("failed to decode "+
+				"header hex: %w", err)
+		}
+
+		_, err = tempFile.Write(hdrBytes)
+		if err != nil {
+			return nil, cleanup, fmt.Errorf("failed to write "+
+				"header data: %w", err)
+		}
+	}
+
+	// Sync to ensure all data is written to disk.
+	if err := tempFile.Sync(); err != nil {
+		return nil, cleanup, fmt.Errorf("failed to sync file: %w", err)
+	}
+
+	// Reset file position to beginning for reading.
+	if _, err := tempFile.Seek(0, 0); err != nil {
+		return nil, cleanup, fmt.Errorf("failed to reset file "+
+			"position: %w", err)
+	}
+
+	return tempFile, cleanup, nil
+}
+
+// constructBlkHdr constructs a block header from a hex string and height.
+func constructBlkHdr(blockHeaderHex string,
+	height uint32) (*BlockHeader, error) {
+
+	buff, err := hex.DecodeString(blockHeaderHex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode block header hex: %v",
+			err)
+	}
+	reader := bytes.NewReader(buff)
+	bHExpected := NewBlockHeader()
+	bHExpected.Deserialize(reader, height)
+	bH, ok := bHExpected.(*BlockHeader)
+	if !ok {
+		return nil, errors.New("failed to assert *BlockHeader type")
+	}
+	return bH, nil
+}
+
+// constructFilterHdr constructs a filter header from a hex string and height.
+func constructFilterHdr(filterHeaderHex string,
+	height uint32) (*FilterHeader, error) {
+
+	buff, err := hex.DecodeString(filterHeaderHex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode filter header hex: %v",
+			err)
+	}
+	reader := bytes.NewReader(buff)
+	fHExpected := NewFilterHeader()
+	fHExpected.Deserialize(reader, height)
+	fH, ok := fHExpected.(*FilterHeader)
+	if !ok {
+		return nil, errors.New("failed to assert *FilterHeader type")
+	}
+	return fH, nil
+}
```

Comment:
Not sure I can review 8k lines of test code in great detail. But looks quite thorough, nice work!
My approach for manually testing this PR is: Adapt https://github.com/guggero/block-dn to also create the files in the format expected by this PR, then do a full mainnet pre-load cycle.
Going to report back with the results of that later.